### PR TITLE
Multilogin

### PR DESCRIPTION
**Pull request by borisyankov**
* Support multiple account logins
* Add NavigationExperimental to handle login flow and main screen
* Add new screen - Account List
* All accounts entered are persisted and can be deleted
* Handling of partial logins - during login process (have only realm, or only realm+password)

### CODE REVIEW

**Comment by smarx** [comment]
*Automated message from Dropbox CLA bot*

@borisyankov, it looks like you've already signed the Dropbox CLA. Thanks!

**Comment by timabbott** [comment]
Awesome, I'm super excited about this :)

File: src/accountlist/accountlistReducers.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import {
+  REALM_ADD,
+  LOGIN_SUCCESS,
+  LOGOUT,
+  ACCOUNT_REMOVE,
+} from '../account/accountActions';
+
+const initialState = fromJS([]);
+
+export default (state = initialState, action) => {
```
Comment (neerajwahi) [inline src/accountlist/accountlistReducers.js]: I don't think I like the pattern of having the first account in the list be the active account. It's strange to have to remove an account and add it back to the front of the list

Why not store the list of accounts as a map and have `activeAccountId` as part of the state (like it used to be before the auth refactor)?

File: src/common/Screen.js [hunk 2]
```diff
@@ -0,0 +1,48 @@
+import React from 'react';
+import {
+  // BackAndroid,
```
Comment (neerajwahi) [inline src/common/Screen.js]: I'd remove this

File: src/common/index.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+export { default as Avatar } from './Avatar';
```
Comment (neerajwahi) [inline src/common/index.js]: Where is this file used?

I think we should avoid the pattern of moving imports to another file for convenience

File: src/main/MainScreen.js [hunk 4]
```diff
@@ -9,30 +9,16 @@ import Drawer from 'react-native-drawer';
...
-} from '../nav/navActions';
+import { sendInitialGetUsers } from '../userlist/userListActions';
+import { appActivity } from '../account/appActions';
+import { sendGetMessages, sendSetMessages } from '../stream/streamActions';
+import { getEvents } from '../events/eventActions';
+import { openStreamSidebar, closeStreamSidebar } from '../nav/navActions';
 
 import StreamView from '../stream/StreamView';
-import NavBar from '../nav/NavBar';
+import MainNavBar from '../nav/MainNavBar';
```
Comment (neerajwahi) [inline src/main/MainScreen.js]: Why rename this?

File: src/message/MessageView.js [hunk 5]
```diff
@@ -34,7 +34,7 @@ const styles = StyleSheet.create({
     fontSize: 15,
   },
   messageTime: {
-    color: '#999',
+    color: BORDER_COLOR,
```
Comment (neerajwahi) [inline src/message/MessageView.js]: This is much lighter than it used to be. And I don't think its style should be related `BORDER_COLOR` in any case

File: src/nav/MainNavBar.js [hunk 6]
```diff
@@ -18,30 +16,25 @@ const styles = StyleSheet.create({
   },
   navBar: {
     flexDirection: 'row',
-    height: NAV_BAR_HEIGHT + STATUS_BAR_HEIGHT,
-    backgroundColor: '#fff',
-    paddingTop: STATUS_BAR_HEIGHT,
-    alignItems: 'center',
+    backgroundColor: BRAND_COLOR,
```
Comment (neerajwahi) [inline src/nav/MainNavBar.js]: We should add back a lot of the old styles for the nav bar. It's now much smaller than standard for iOS and alignment of the heading is no longer centered

In general, I think we should put styling changes in separate PRs from functionality changes

File: src/nav/navActions.js [hunk 7]
```diff
@@ -1,6 +1,30 @@
...
+export const pushRoute = (route: Object) => ({
+  type: PUSH_ROUTE,
+  route,
+});
+
+export const popRoute = () => ({
+  type: POP_ROUTE,
+});
+
+export const changeTab = (index: number) => ({
```
Comment (neerajwahi) [inline src/nav/navActions.js]: I don't see this used anywhere

File: src/nav/navReducers.js [hunk 8]
```diff
@@ -29,9 +45,22 @@ const reducer = (state = initialState, action) => {
         ...state,
         opened: false,
       };
+    case INIT_ROUTES:
+      return NavigationStateUtils.reset(
+        state,
+        action.routes.map(route => ({ key: route }))
+      );
+    case PUSH_ROUTE:
+      if (state.routes[state.index].key === (action.route && action.route.key)) return state;
```
Comment (neerajwahi) [inline src/nav/navReducers.js]: This line seems really long

File: src/reducers.js [hunk 9]
```diff
@@ -1,22 +1,24 @@
 import { combineReducers } from 'redux';
+import accountlist from './accountlist/accountlistReducers';
 import app from './account/appReducers';
-import auth from './account/authReducers';
-import user from './account/userReducers';
-import userlist from './userlist/userListReducers';
-import stream from './stream/streamReducers';
+// import auth from './account/authReducers';
```
Comment (neerajwahi) [inline src/reducers.js]: We should remove unneeded lines that are commented out

File: src/reducers.js [hunk 10]
```diff
@@ -1,22 +1,24 @@
...
+import userlist from './userlist/userListReducers';
 
 export default combineReducers({
+  accountlist,
   app,
-  auth,
-  user,
-  userlist,
-  stream,
+//  auth,
```
Comment (neerajwahi) [inline src/reducers.js]: Ditto above

File: src/start/RealmScreen.js [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+import { bindActionCreators } from 'redux';
+import { connect } from 'react-redux';
+
+import { styles, Screen, ErrorMsg, Button } from '../common';
+import { getAuthBackends } from '../api/apiClient';
+import config from '../config';
+import { realmAdd } from '../account/accountActions';
+import AuthBackendPicker from './AuthBackendPicker';
+
+type Props = {
```
Comment (neerajwahi) [inline src/start/RealmScreen.js]: We should start adding `// @flow` to the beginning of files when we can. Otherwise they don't get typechecked

File: src/start/PasswordAuthScreen.js [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+  }
+
+  tryPasswordLogin = async () => {
+    const { auth } = this.props;
+    const { email, password } = this.state;
+
+    this.setState({ progress: true, error: '' });
+
+    try {
+      const apiKey = await fetchApiKey(auth, email, password);
```
Comment (neerajwahi) [inline src/start/PasswordAuthScreen.js]: Why refactor it this way? (instead of just calling a single action and letting it flow through Redux)

File: src/start/RealmScreen.js [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+    };
+  }
+
+  tryRealm = async () => {
+    const { realm } = this.state;
+
+    this.setState({ progress: true, error: '' });
+
+    try {
+      const authBackends = await getAuthBackends({ realm });
```
Comment (neerajwahi) [inline src/start/RealmScreen.js]: Same as above. What are the advantages of doing it this way? I think I prefer the way this used to work

File: src/common/Touchable.js [hunk 14]
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
```
Comment (neerajwahi) [inline src/common/Touchable.js]: Thanks for moving this out into a common component! Makes it much easier to maintain a consistent style :)

File: src/accountlist/__tests__/getActiveAccount-test.js [hunk 15]
```diff
@@ -0,0 +1,14 @@
+import { fromJS } from 'immutable';
+import { getActiveAccount } from '../accountlistSelectors';
+
+it('when no accounts exist, returns undefined', () => {
```
Comment (borisyankov) [inline src/accountlist/__tests__/getActiveAccount-test.js]: Yep, agree.

File: src/common/__tests__/__snapshots__/Button-test.js.snap [hunk 16]
```diff
@@ -2,19 +2,17 @@ exports[`test renders correctly 1`] = `
 <View
```
Comment (borisyankov) [inline src/common/__tests__/__snapshots__/Button-test.js.snap]: Will remove it, and we should investigate where snapshot testing makes sense. It seems it does wth redux and other use-cases and maybe for visual components once stuff settles down.

File: src/common/index.js (hunk 3 above)
Comment (borisyankov) [inline src/common/index.js]: This enables the pattern `import { one, two, three } from './common'`

File: src/main/MainScreen.js (hunk 4 above)
Comment (borisyankov) [inline src/main/MainScreen.js]: Because now there are two nav bars, and possibly more in future.

File: src/message/MessageView.js (hunk 5 above)
Comment (borisyankov) [inline src/message/MessageView.js]: Yeah, not related. Changed by mistake I guess.

File: src/accountlist/accountlistReducers.js (hunk 1 above)
Comment (borisyankov) [inline src/accountlist/accountlistReducers.js]: I am making the most recent account bubble up for UI purposes too - most recent is on top.
If it wasn't for that, it would have not been natural to do so, just to indicate which one is active.

File: src/start/RealmScreen.js (hunk 13 above)
Comment (borisyankov) [inline src/start/RealmScreen.js]: https://github.com/reactjs/redux/issues/1287

File: src/accountlist/accountlistReducers.js (hunk 1 above)
Comment (neerajwahi) [inline src/accountlist/accountlistReducers.js]: Oh ok. How about adding a timestamp then and sorting in the UI component? I'm not a big fan of coupling UI (such as 'sorted by most recent') with the data model.

I think the logic for `REALM_ADD` and `LOGIN_SUCCESS` below is too complex.

File: src/accountlist/accountlistSelectors.js [hunk 17]
```diff
@@ -0,0 +1,22 @@
+import { fromJS } from 'immutable';
+
+export const getActiveAccount = (state) =>
+  state.accountlist.get(0);
+
+export const getAuth = (state) => {
+  const account = getActiveAccount(state);
+
+  if (!account) {
+    return fromJS({
```
Comment (neerajwahi) [inline src/accountlist/accountlistSelectors.js]: Why use `fromJS` instead of explicit constructors (such as `Immutable.Record` or `Immutable.Map`)?

File: src/common/index.js (hunk 3 above)
Comment (neerajwahi) [inline src/common/index.js]: 👍

File: src/nav/routingSelectors.js [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+  const routeList = [];
+
+  if (accounts.size > 1) routeList.push('accountlist');
+
+  routeList.push('realm');
+
+  if (activeAccount) {
+    if (activeAccount.get('email')) {
+      routeList.push('password');
+    } else if (activeAccount.get('todo')) {
```
Comment (neerajwahi) [inline src/nav/routingSelectors.js]: Let's get rid of the 'todos' and add this logic back when ready (just so that anyone looking at this commit in the future isn't confused).

File: src/start/PasswordAuthScreen.js [hunk 19]
```diff
@@ -0,0 +1,92 @@
...
+      email: props.email || config.defaultLoginEmail,
+      password: props.password || config.defaultLoginPassword,
+    };
+  }
+
+  tryPasswordLogin = async () => {
+    const { auth } = this.props;
+    const { email, password } = this.state;
+
+    this.setState({ progress: true, error: '' });
```
Comment (neerajwahi) [inline src/start/PasswordAuthScreen.js]: Can we make `error` null here? I think it's clearer if the no error case isn't an empty string

File: src/start/RealmScreen.js (hunk 13 above)
Comment (neerajwahi) [inline src/start/RealmScreen.js]: Thanks for the link. I'm not sure I agree that the state should live locally in this component in this case, but we can chat about that later

File: src/accountlist/accountlistSelectors.js (hunk 17 above)
Comment (borisyankov) [inline src/accountlist/accountlistSelectors.js]: No special reason, they are identical. Often is shorter and does deep conversion, not just one level. I am open to changing it at later stage.

File: src/accountlist/accountlistReducers.js (hunk 1 above)
Comment (borisyankov) [inline src/accountlist/accountlistReducers.js]: Let's reconsider this after we discuss with Tim and others if this behavior is ideal, and what we might want to add to the login flow (maybe not much). I definitely prefer the array, if no timestamp is ever used, as the state is not a db and using the array's property to be ordered is a valid approach, IMHO.

Comment (neerajwahi) [inline src/accountlist/accountlistReducers.js]: Sure, we can discuss the behavior and what we want to add, but in any case I'd like to change this. I don't like coupling the ordering of the list in the UI to the state reducer logic.

File: src/accountlist/accountlistSelectors.js (hunk 17 above)
Comment (neerajwahi) [inline src/accountlist/accountlistSelectors.js]: Yeah I think we should avoid deep conversion. For example, that allows us to keep messages in an Immutable array, but the elements themselves can be the raw objects we retrieve from the server (which is more performant)

File: src/accountlist/accountlistReducers.js (hunk 1 above)
Comment (neerajwahi) [inline src/accountlist/accountlistReducers.js]: I'm open to changing my mind of course. I just don't understand the benefit of having a sorted array in the state (as opposed to sorting at the UI level). Could you elaborate a little on that?
